# _Tasca 2, Sprint 3_ *Data Structures* 

## Exercise 1

Create a list that groups the months of the year into quarters (Q1: January, February and March, Q2: April, May, June ...), that is, a list with 4 lists inside.

In [1]:
q1 = ['January', 'February', 'March']
q2 = ['April', 'May', 'June']
q3 = ['July', 'August', 'September']
q4 = ['October', 'November', 'December']

In [2]:
months = [q1, q2, q3, q4]
months

[['January', 'February', 'March'],
 ['April', 'May', 'June'],
 ['July', 'August', 'September'],
 ['October', 'November', 'December']]

## Exercise 2

Create a code that allows you to access:

    The second month of the first quarter
    The months of the first quarter
    September and October

In [3]:
t1 = q1[1]
print('The second month of the first quarter is: ' + t1)

The second month of the first quarter is: February


In [4]:
t2 = q1
print('The months of the first quarter are: ' + t2[0] + ', ' + t2[1] + ' and ' + t2[2] )

The months of the first quarter are: January, February and March


In [5]:
t3 = [q3[2], q4[0]]
print(t3[0] + ' and ' + t3[1])

September and October


## Exercise 3

Create a list of disordered numbers and answer the following questions:

    1. How many numbers are there?
    2. How many times does the number 3 appear
    3. How many times do the numbers 3 and 4 appear?
    4. What is the largest number?
    5. What are the 3 smallest numbers?
    6. What is the range of this list?

In [6]:
import random
random.seed(15)
numbers = random.sample(range(1, 50), 20)
print(numbers)

[14, 1, 34, 3, 11, 16, 2, 4, 10, 24, 44, 8, 22, 30, 23, 18, 26, 17, 12, 38]


In [7]:
print ('Total numbers: ' + str(len(numbers)))
print ('Number 3: ' + str(numbers.count(3)))
print ('Number 3 and 4: ' + str(numbers.count(3)) + ' and ' + str(numbers.count(3)))
print ('The largest number: ' + str(max(numbers)))
print ('The 3 smallest numbers: ' + str(sorted(numbers)[:3]))
print ('The range of the list: ' + str(min(numbers)) + ' to ' + str(max(numbers)))

Total numbers: 20
Number 3: 1
Number 3 and 4: 1 and 1
The largest number: 44
The 3 smallest numbers: [1, 2, 3]
The range of the list: 1 to 44


**Answers to the questions for Excercise 3:**

    1. There are 20 numbers in the list
    2. Number 3 appears 1 times
    3. Number 3 and 4 appears 1 and 1 times
    4. The largest number is 44
    5. The 3 smallest numbers are 1, 2, and 3
    6. The range of the list is [1, 44]

##  Exercise 4

Create a dictionary as follows and answer the questions:

purchase = {"Apples": {"Qty": 5, "€": 0.42}, "Pears": {"Qty": 3, "€": 0.66}}

    1. Add some more fruit
    2. How much did the pears cost in total?
    3. How many fruits did we buy in total?
    4. What is the most expensive fruit?

In [8]:
purchase = {"Apples": {"Qty": 5, "€": 0.42}, "Pears": {"Qty": 3, "€": 0.66}, "Bananas": {"Qty": 7, "€": 0.36}, "Peaches": {"Qty": 2, "€": 0.76}, "Mangoes": {"Qty": 2, "€": 1.58}, "Pineapples": {"Qty": 0, "€": 1.38}}
purchase

{'Apples': {'Qty': 5, '€': 0.42},
 'Pears': {'Qty': 3, '€': 0.66},
 'Bananas': {'Qty': 7, '€': 0.36},
 'Peaches': {'Qty': 2, '€': 0.76},
 'Mangoes': {'Qty': 2, '€': 1.58},
 'Pineapples': {'Qty': 0, '€': 1.38}}

In [9]:
fruit = "Pears" # fruit for which we want to know the total cost
cost_fruit = purchase[fruit]["Qty"] * purchase[fruit]["€"]
print('Total cost of ' + fruit + ': ' + str(cost_fruit))

Total cost of Pears: 1.98


In [10]:
# Let's get the list of fruits in the purchase list
fruitlist = []
fruitlist = [item[:] for item in purchase]
print(fruitlist)

# The following calculates the quantity of each fruit in the list and adds them
totalfruits = 0
for fruit in fruitlist:
    totalfruits = totalfruits + purchase[fruit]["Qty"]
    print(fruit + ': ' + str(purchase[fruit]["Qty"]))

print('Fruits bought in total: ' + str(totalfruits))

['Apples', 'Pears', 'Bananas', 'Peaches', 'Mangoes', 'Pineapples']
Apples: 5
Pears: 3
Bananas: 7
Peaches: 2
Mangoes: 2
Pineapples: 0
Fruits bought in total: 19


In [11]:
price = []
for fruit in fruitlist:
    cost = [purchase[fruit]["€"]]
    #print(cost)
    price.append(cost)
print(price)

[[0.42], [0.66], [0.36], [0.76], [1.58], [1.38]]


In [12]:
x = max(price)
for fruit in fruitlist:
    if x[0] == purchase[fruit]["€"]:
        print('Expensive fruit: ' + fruit)

Expensive fruit: Mangoes


**Answers to the questions for Excercise 4:**

    1. Added Bananas, Peaches, Mangoes and Pineapples.
    2. Pears cost 1.98 € in total.
    3. We bought 19 fruits in total.
    4. The most expensive fruit is Mango. 